mptcp: mpc: split ACK and push

matttbe opened the pull request:
This test recently failed with this error when validating @pabeni's series:

    v1_bind_tcpfallback_flagH_3rd_ack.pkt:25: error handling packet: live packet payload: expected 0 bytes vs actual 1000 bytes

https://github.com/multipath-tcp/mptcp_net-next/actions/runs/9813475267

Likely because Packetdrill was too slow to inject the packet with the data. Then sending only the ACK alone should help.

matttbe committed:
mptcp: mpc: split ACK and push

The following test recently failed with this error:

  v1_bind_tcpfallback_flagH_3rd_ack.pkt:25: error handling packet: live packet payload: expected 0 bytes vs actual 1000 bytes

Likely because Packetdrill was too slow to inject the packet with the
data. Then sending only the ACK alone should help. Note that we had a
similar issue with ACK + ADD_ADDR.

All other tests doing the similar thing have been modified as well.

Signed-off-by: Matthieu Baerts (NGI0) <matttbe@kernel.org>

dcaratti commented on gtests/net/mptcp/mp_capable/v1_bind_tcpfallback_flagH_3rd_ack.pkt:
```diff
@@ -21,6 +21,7 @@
 
 +0     write(4, ..., 1000) = 1000
 +0       >  P.  1:1001(1000)  ack 1
++0       <   .  1:1(0)        ack 1001  win 257
```
this is caused by TCP retransmitting too fast / too aggressively the packet in line 23. Can this be fixed by adjusting the estimated RTT like done elsewhere?

matttbe replied:
> this is caused by TCP retransmitting too fast / too aggressively the packet in line 23.

@dcaratti: Yes, that's what I assumed, because the ACK from the next packet -- also including data -- has been injected with a too long delay.

> Can this be fixed by adjusting the estimated RTT like done elsewhere?

Hard to tell: the only delay estimation the kernel had been able to compute is the delay between the SYN+ACK and the ACK: `0.01`. We can switch to `0.1` instead, but I suppose the delay here was caused by the fact the injected packet was not just a "simple" ACK, but carried data. That's why I added this dedicated ACK instead.

Most tests add a delay of 0.01. I suggest to check if we still have such instabilities after this modification, and if yes, we can switch all of them to 0.1, not just this one. WDYT?

dcaratti replied:
I can't tell how much additional time is needed by packetdrill to fill packet with data, but I don't believe it can be 
comparable to 10ms on a modern system. I'm more inclined to think that packetdrill process is not responsive enough  because of scheduler settings on the test VM. Maybe we can find some settings and re-try?

matttbe committed:
mptcp: no PUSH flag with no data

Some injected packets had a PUSH flag, but not carrying any data.

The kernel handles that correctly, but it's not something we should
have. So we can remove it.

Signed-off-by: Matthieu Baerts (NGI0) <matttbe@kernel.org>

dcaratti commented:
LGTM, thnaks!
